src/components/ApiHistoryTable.tsx

```diff
@@ -0,0 +1,160 @@
+      <a
+        key={version}
+        href={change['pr-url']}
+        target="_blank"
+        rel="noopener noreferrer"
+      >
+        {/* Semver shenanigans, feature backported to both ^7.1.0 and ^6.3.0 would not be present in 7.0.0 */}
+        <pre>
+          {release === version ? '>=' : '^'}
+          {version}
+        </pre>
+      </a>
+    );
+  });
+
+  let changesJsx: JSX.Element | undefined;
+
+  if ('description' in change) {
+    changesJsx = (
+      <ReactMarkdown
+        allowedElements={['a', 'code', 'em', 'p', 'pre', 'strong']}
+      >
+        {change['description']}
+      </ReactMarkdown>
+    );
+  } else {
+    changesJsx = (
+      <pre
+        key={key}
+        className={styles[type.toLowerCase().split(' ').join('-')]}
+      >
+        {type}
+      </pre>
+    );
+  }
+
+  if ('breaking-changes-header' in change) {
+    changesJsx = (
+      <a
+        href={
+          // ? Is this okay? Is there an existing function that should be used here to generate the URL?
+          '/docs/latest/breaking-changes#' + change['breaking-changes-header']
+        }
+        target="_blank"
+        rel="noopener noreferrer"
+      >
+        {changesJsx}
+      </a>
+    );
+  }
+
+  return (
+    <tr key={key}>
+      <td>
+        {formattedVersions.length !== 0 ? formattedVersions : <pre>None</pre>}
+      </td>
+      <td>{changesJsx}</td>
+    </tr>
+  );
+}
+
+const ApiHistoryTable = (props: ApiHistoryTableProps) => {
+  const { apiHistoryJson, prReleaseVersionsJson } = props;
+
+  const apiHistory = JSON.parse(apiHistoryJson) as ApiHistory;
+  const prReleaseVersions = JSON.parse(
+    prReleaseVersionsJson
+  ) as PrReleaseVersionsContainer;
+
+  // ? Maybe this is too much abstraction?
+  // ? `added` and `deprecated` go on the ends, but is there any point in sorting `changes`?
+  //  How do you sort them? Given change X and change Y, X could have been merged into main before Y,
+  //  but Y could have been backported to a stable version before X.
+  // ? Is it possible for neither the release or any of the backports for a change to be in a released stable version?
+  //  In that case, there would be no point in generating a row for that change.
+  const apiHistoryChangeRows = [
+    ...(apiHistory.deprecated?.map((deprecated) => {
+      const prNumber = Number(deprecated['pr-url'].split('/').at(-1));
+      return generateTableRow(
+        prReleaseVersions[prNumber],
+        Change.DEPRECATED,
+        deprecated
+      );
+    }) ?? []),
+    ...(apiHistory.changes?.map((change) => {
+      const prNumber = Number(change['pr-url'].split('/').at(-1));
+      return generateTableRow(
+        prReleaseVersions[prNumber],
+        Change.CHANGED,
+        change
+      );
+    }) ?? []),
+    ...(apiHistory.added?.map((added) => {
+      const prNumber = Number(added['pr-url'].split('/').at(-1));
+      return generateTableRow(prReleaseVersions[prNumber], Change.ADDED, added);
+    }) ?? []),
+  ];
```

Comment:
```suggestion
  const generateChangeRows = (changeType, changes) => {
    return (changes ?? []).map((change) => {
      const prNumber = Number(change['pr-url'].split('/').at(-1));
      return generateTableRow(
        prReleaseVersions[prNumber],
        changeType,
        change
      );
    });
  };

  const apiHistoryChangeRows = [
    ...generateChangeRows(Change.DEPRECATED, apiHistory.deprecated),
    ...generateChangeRows(Change.CHANGED, apiHistory.changes),
    ...generateChangeRows(Change.ADDED, apiHistory.added),
  ];
```

IMO this reads a lot cleaner if you pull out the common code. Type annotations left as an exercise for the reader. 🙂

Reply:
Looks great, good idea 👍